File: leetcode/verifying-an-alien-dictionary.py
from typing import List

class Solution:
    def converLetterToNumber(self, letter: str) -> int:
        return ord(letter) - ord('a')

    def isWordAGreaterThanWordB(self, a: str, b: str, letterOrder: List[int]) -> bool:
        for i in range(min(len(a),len(b))):
            letterOrderA = letterOrder[self.converLetterToNumber(a[i])]
            letterOrderB = letterOrder[self.converLetterToNumber(b[i])]
            if(letterOrderA > letterOrderB):
                return True
            if(letterOrderA < letterOrderB):
                return False 

        return len(a) > len(b) 

    def isAlienSorted(self, words: List[str], order: str) -> bool:
        letterOrder = [0 for _ in range(len(order))]
        for i in range(len(order)):
            letterOrder[self.converLetterToNumber(order[i])] = i

        last = words[0]
        for word in words:
            if(self.isWordAGreaterThanWordB(last,word, letterOrder)):
                return False
            last = word

        return True 

def main():
    words = ["hello","leetcode"]
    order = "hlabcdefgijkmnopqrstuvwxyz"
    print(Solution().isAlienSorted(words,order))
    words = ["word","world","row"]
    order = "worldabcefghijkmnpqstuvxyz"
    print(Solution().isAlienSorted(words,order))

if __name__ == "__main__":
    main()

File: leetcode/verifying-an-alien-dictionary.cpp
#include <bits/stdc++.h>
using namespace std;

#define pb push_back
#define eb emplace_back
#define mk make_pair
#define fi first
#define se second
#define cc(x) cout << #x << " = " << x << endl
#define ok cout << "ok" << endl
#define endl '\n'

typedef long long ll;
typedef pair<int, int> ii;
const int INF = 0x3f3f3f3f;
const double PI = acos(-1.0);

class Solution {
public:
  bool isWordAAlienGreaterThanB(string &a, string &b,
                                vector<int> &letterOrder) {
    for (int letter = 0; letter < min(a.size(), b.size()); letter++) {
      int posALetter = letterOrder[a[letter] - 'a'];
      int posBLetter = letterOrder[b[letter] - 'a'];
      if (posALetter > posBLetter) {
        return true;
      }
      if (posALetter < posBLetter) {
        return false;
      }
    }
    return a.size() > b.size();
  }

  bool isAlienSorted(vector<string> &words, string order) {
    vector<int> letterOrder(27, 0);
    for (int i = 0; i < order.size(); i++) {
      letterOrder[order[i] - 'a'] = i;
    }

    string last = words[0];

    for (string word : words) {
      if (isWordAAlienGreaterThanB(last, word, letterOrder)) {
        return false;
      }
      last = word;
    }

    return true;
  }
};

int main() {
  ios_base::sync_with_stdio(false);
  vector<string> words = {"word", "world", "row"};
  string order = "worldabcefghijkmnpqstuvxyz";
  Solution *solution = new Solution();
  bool ans = solution->isAlienSorted(words, order);
  cout << (ans ? "true" : "false") << endl;
  words = {"hello", "leetcode"};
  order = "hlabcdefgijkmnopqrstuvwxyz";
  ans = solution->isAlienSorted(words, order);
  cout << (ans ? "true" : "false") << endl;
  return 0;
}
